packages/client/wallets/smart-wallet/src/blockchain/wallets/SendTransactionService.ts

```diff
@@ -0,0 +1,191 @@
+import type { SmartAccountClient } from "permissionless";
+import type { EntryPoint } from "permissionless/_types/types";
+import type { SmartAccount } from "permissionless/accounts";
+import {
+    type Abi,
+    type Address,
+    BaseError,
+    type Chain,
+    ContractFunctionRevertedError,
+    type Hex,
+    type PublicClient,
+    type TransactionReceipt,
+    type Transport,
+} from "viem";
+
+import { CrossmintSDKError, WalletErrorCode } from "@crossmint/client-sdk-base";
+
+export type TransactionServiceTransactionRequest = {
+    address: Address;
+    abi: Abi;
+    functionName: string;
+    args: any;
+    value?: bigint;
+};
+
+export type SendTransactionFailureStage = "simulation" | "send" | "confirmation" | "execution";
+
+/**
+ * Error thrown when a transaction fails to send.
+ * @param viemError The error thrown by the viem client. See https://viem.sh/docs/glossary/errors.html
+ */
+export class EVMSendTransactionError extends CrossmintSDKError {
+    constructor(
+        message: string,
+        public readonly viemError: BaseError,
+        public readonly stage: SendTransactionFailureStage,
+        code = WalletErrorCode.SEND_TRANSACTION_FAILED
+    ) {
+        super(message, code);
+    }
+}
+
+/**
+ * Error thrown when a transaction fails due to a contract execution error.
+ * @param viemError The error thrown by the viem client. See https://viem.sh/docs/glossary/errors.html
+ * @param revertError The revert error from viem containing the reason for the revert.
+ * @param revertError.reason The reason for the revert.
+ * @param revertError.data The decoded revert error data.
+ * @example
+ * try {
+ *   await wallet.sendTransaction({
+ *     address: contractAddress,
+ *     abi,
+ *     functionName: "mintNFT",
+ *     args: [recipientAddress],
+ *   });
+ * } catch (e) {
+ *   if (e instanceof SendTransactionExecutionRevertedError) {
+ *     alert(`Transaction reverted: ${e.revertError.reason}`);
+ *   }
+ *   throw e;
+ * }
+ */
+export class EVMSendTransactionExecutionRevertedError extends EVMSendTransactionError {
+    constructor(
+        message: string,
+        public readonly viemError: BaseError,
+        revertError: ContractFunctionRevertedError,
```

Comment:
Now that we've exposed it's useful properties here, I'm not sure there's a need for `revertError` to be included.

Reply:
It's not included, it's best to pass it to the constructor so that the reassignment logic can be contained in the error

Reply:
Ahh, ofc. misread this